src/stream/src/executor/aggregation/state_cache/mod.rs

```diff
@@ -12,262 +12,135 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use std::collections::BTreeMap;
-
-use risingwave_common::array::Op;
 use risingwave_common::types::{Datum, DatumRef};
 use smallvec::SmallVec;
 
-use self::cache::OrderedCache;
 use super::minput::StateCacheInputBatch;
+use crate::common::cache::{StateCache, StateCacheFiller};
 
 pub mod array_agg;
-mod cache;
 pub mod extreme;
 pub mod string_agg;
 
 /// Cache key type.
 pub type CacheKey = Vec<u8>;
 
-/// Trait that defines the interface of state table cache.
-pub trait StateCache: Send + Sync + 'static {
+/// Trait that defines the interface of state table cache for stateful streaming agg.
+pub trait AggStateCache {
     /// Check if the cache is synced with state table.
     fn is_synced(&self) -> bool;
 
     /// Apply a batch of updates to the cache.
     fn apply_batch(&mut self, batch: StateCacheInputBatch<'_>);
 
     /// Begin syncing the cache with state table.
-    fn begin_syncing(&mut self) -> StateCacheFiller<'_>;
+    fn begin_syncing(&mut self) -> Box<dyn AggStateCacheFiller + Send + Sync + '_>;
 
     /// Get the aggregation output.
     fn get_output(&self) -> Datum;
 }
 
-/// Cache maintenance interface.
-/// Note that this trait must be private, so that only [`StateCacheFiller`] can use it.
-trait StateCacheMaintain: Send + Sync + 'static {
+/// Trait that defines agg state cache syncing interface.
+pub trait AggStateCacheFiller {
+    /// Get the capacity of the cache to be filled. `None` means unlimited.
+    fn capacity(&self) -> Option<usize>;
+
     /// Insert an entry to the cache without checking row count, capacity, key order, etc.
     /// Just insert into the inner cache structure, e.g. `BTreeMap`.
-    fn insert_unchecked(&mut self, key: CacheKey, value: SmallVec<[DatumRef<'_>; 2]>);
+    fn append(&mut self, key: CacheKey, value: SmallVec<[DatumRef<'_>; 2]>);
 
     /// Mark the cache as synced.
-    fn set_synced(&mut self);
-}
-
-/// A temporary handle for filling the state cache.
-/// The state cache will be marked as synced automatically when this handle is dropped.
-pub struct StateCacheFiller<'a> {
-    capacity: usize,
-    cache: &'a mut dyn StateCacheMaintain,
-}
-
-impl<'a> StateCacheFiller<'a> {
-    /// Get the capacity of the cache to be filled.
-    pub fn capacity(&self) -> usize {
-        self.capacity
-    }
-
-    /// Insert an entry to the cache.
-    pub fn insert(&mut self, key: CacheKey, value: SmallVec<[DatumRef<'_>; 2]>) {
-        self.cache.insert_unchecked(key, value)
-    }
-
-    /// Finish the cache filling process.
-    /// Must be called after inserting all entries to mark the cache as synced.
-    pub fn finish(self) {
-        self.cache.set_synced();
-    }
+    fn finish(self: Box<Self>);
```

Comment:
Is this `Box` necessary?

Reply:
Yes, because cache filler is expected to be returned by `begin_syncing` as `Box<dyn XxxFiller>`, and the statement `filler.finish()` cannot statically determine the size of `self`.